File: 1248-count-number-of-nice-subarrays/1248-count-number-of-nice-subarrays.java
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {

        // convert even→0, odd→1
        for(int i = 0; i < nums.length; i++){
            nums[i] = nums[i] % 2;
        }

        HashMap<Integer, Integer> map = new HashMap<>();

        map.put(0, 1); // important

        int prefixSum = 0;
        int count = 0;

        for(int num : nums){

            prefixSum += num;

            if(map.containsKey(prefixSum - k)){
                count += map.get(prefixSum - k);
            }

            map.put(prefixSum, map.getOrDefault(prefixSum, 0) + 1);
        }

        return count;
    }
}

File: 1248-count-number-of-nice-subarrays/1248-count-number-of-nice-subarrays.py
class Solution:
    def numberOfSubarrays(self, nums: list[int], k: int) -> int:
        def mostK(k : int) -> int:
            l = res = odd = 0

            for i in range(len(nums)):
                odd += nums[i] % 2

                while odd > k:
                    odd -= nums[l] % 2
                    l += 1
                res += i - l + 1
            return res
        return mostK(k) - mostK(k-1)
